### PR TITLE
Check google provider version in GCSToS3Operator before provide match_glob param

### LINKED ISSUE #32920: GCSToS3Operator is providing an unexpected argument to GCSHook.list
### Apache Airflow version

2.6.3

### What happened

https://github.com/apache/airflow/blob/d800c1bc3967265280116a05d1855a4da0e1ba10/airflow/providers/amazon/aws/transfers/gcs_to_s3.py#L148-L150

This line in `GCSToS3Operator` is currently broken on Airflow 2.6.3:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/airflow/providers/amazon/aws/transfers/gcs_to_s3.py", line 148, in execute
    files = hook.list(
TypeError: GCSHook.list() got an unexpected keyword argument 'match_glob'
```

The call signature for `GCSHook.list` does not have a `match_glob` argument on Airflow 2.6.3
https://github.com/apache/airflow/blob/eb24742d5300d2d87b17b4bcd67f639dbafd9818/airflow/providers/google/cloud/hooks/gcs.py#L699

However it does on the `main` branch:
https://github.com/apache/airflow/blob/0924389a877c5461733ef8a048e860b951d81a56/airflow/providers/google/cloud/hooks/gcs.py#L702-L710

It appears that `GCSToS3Operator` jumped the gun on using the `match_glob` 

### What you think should happen instead

_No response_

### How to reproduce

Create a task that uses `airflow.providers.amazon.aws.transfers.gcs_to_s3.GCSToS3Operator`. Execute the task.

### Operating System

Debian GNU/Linux 11 (bullseye)

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==8.3.1
apache-airflow-providers-celery==3.2.1
apache-airflow-providers-cncf-kubernetes==7.3.0
apache-airflow-providers-common-sql==1.6.0
apache-airflow-providers-datadog==3.3.1
apache-airflow-providers-dbt-cloud==3.2.2
apache-airflow-providers-elasticsearch==4.5.1
apache-airflow-providers-ftp==3.4.2
apache-airflow-providers-google==10.0.0
apache-airflow-providers-http==4.5.0
apache-airflow-providers-imap==3.2.2
apache-airflow-providers-microsoft-azure==6.2.1
apache-airflow-providers-postgres==5.5.2
apache-airflow-providers-redis==3.2.1
apache-airflow-providers-sftp==4.4.0
apache-airflow-providers-slack==7.3.1
apache-airflow-providers-sqlite==3.4.2
apache-airflow-providers-ssh==3.7.1

### Deployment

Astronomer

### Deployment details

Using Astro Runtime 8.8.0

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by hussein-awala**
closes: #32920

This PR adds a check for Google provider version in `GCSToS3Operator` before providing `match_glob` to GCS hook, which is not compatible with versions < 10.3.0. Also it raises an exception when trying to init the provider with match_glob param if the google provider version is not compatible.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by eladkal** [comment]
We already have it. It is generated automatically
https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/index.html#cross-provider-package-dependencies

`pip install apache-airflow-providers-amazon[google]`

**Comment by hussein-awala** [comment]
> We already have it. It is generated automatically

yes, I saw it, but we don't specify the min version of google provider compatible with the desired version of aws provider. If we have google provider 10.0.0 installed and we want to install the aws provider, it will not upgrade the google provider version. I just tested it:
```bash
pip install apache-airflow-providers-amazon[google]==8.3.1

...
Requirement already satisfied: apache-airflow-providers-google in /home/hussein.awala/anaconda3/envs/airflow-env/lib/python3.9/site-packages (from apache-airflow-providers-amazon[google]==8.3.1) (10.0.0)
```
But after this PR, the same command should upgrade the google provider version at least to the min supported version. wdyt?

**Comment by eladkal** [comment]
> yes, I saw it, but we don't specify the min version of google provider compatible with the desired version of aws provider. 

This is by design.
If we accept this PR it means all users of Amazon provider will ve forced to use newest version of Google provider just because 1 operator.

If we accept this approch it means we always have to increase version for all providers always. This may lead to many problems for users.

The current way gives more flexibility, it allow users to choose what to use. There are many compatible versions.

We bump depended version only when we must. When there is a really good reason to force everyone to use the new version.

I think the root cause of the specific user who had trouble with versions did not install Airflow with constraints?

**Comment by potiuk** [comment]
Yes. @eladkal is right. We could for example raise Optional provider feature exception if we find that version is not good for that operator with specifying the version of Google provider needed . That would be a better fix

**Comment by hussein-awala** [comment]
> The current way gives more flexibility, it allow users to choose what to use. There are many compatible versions.

I agree, classifying the versions between compatible and not compatible depends on the used operators/hooks.

> I think the root cause of the specific user who had trouble with versions did not install Airflow with constraints?

I don't think that it's related to constraints, it's just a new feature added in the gcs hook, and `GCSToS3Operator`, and installing the latest version of AWS provider with an old version of google providers makes this operator broken.

> We could for example raise Optional provider feature exception if we find that version is not good for that operator with specifying the version of Google provider needed 

Sounds good. I will implement it in the same PR and update the title and the description.

**Comment by hussein-awala** [comment]
@eladkal @potiuk could you check it now?

File: airflow/providers/amazon/aws/transfers/gcs_to_s3.py [hunk 1]
```diff
@@ -31,6 +34,9 @@
     from airflow.utils.context import Context
 
 
+google_provider_version = version("apache-airflow-providers-google")
```
Comment (potiuk) [inline airflow/providers/amazon/aws/transfers/gcs_to_s3.py]: We have to also handle the case where google provider is installed from sources (editable install). Better check is where we do `from apache.airflow.providers.google import __version__`.

It's a relatively new addition - and we have to `try/except ImportError` for it and handle "unknown" version - but it's been addedd in 10.1.0 version of Google, so checking for '< 10.3.0 or unknown` should be OK.

Comment (hussein-awala) [reply airflow/providers/amazon/aws/transfers/gcs_to_s3.py]: It's a good idea, I just implemented it, could you take a second look?

**Review by potiuk — APPROVED** [review]

**Pull request merged**